Prevent conflicts in channel visibility meta box form

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

This PR adds a prefix to the `visibility` field used in the `Channel Visibility` metabox to prevent any conflicts with other plugins or code snippets sending the same field in the `$_POST` request.

It should resolve an issue faced by a user when going through the setup process: https://wordpress.org/support/topic/not-able-to-get-through-setup/


### Detailed test instructions:

1. Edit a product and change the channel visibility value
2. Confirm that the channel visibility value changes for the product

### Changelog entry

> Fix - Prevent conflicts when storing the channel visibility value

nima-karimi committed:
Prevent conflicts for channel visibility meta box form

Adds a prefix to the `visibility` field used in the `Channel Visibility` metabox to prevent any conflicts with other plugins or code snippets sending the same field in the `$_POST` request.

mikkamp reviewed (commented):
The changes look good and it's handling them as expected. Although I triggered the error and it seems it's calling `handle_submission` three times when saving a product.

```
2021-06-21T09:42:01+00:00 ERROR Automattic\WooCommerce\GoogleListingsAndAds\Admin\MetaBox\ChannelVisibilityMetaBox::handle_submission The value of public must be either of [sync-and-show, dont-sync-and-show].
2021-06-21T09:42:01+00:00 ERROR Automattic\WooCommerce\GoogleListingsAndAds\Admin\MetaBox\ChannelVisibilityMetaBox::handle_submission The value of public must be either of [sync-and-show, dont-sync-and-show].
2021-06-21T09:42:01+00:00 ERROR Automattic\WooCommerce\GoogleListingsAndAds\Admin\MetaBox\ChannelVisibilityMetaBox::handle_submission The value of public must be either of [sync-and-show, dont-sync-and-show].
```

Two of those seem to be because it's hooking into both `woocommerce_update_product` and `woocommerce_process_product_meta`, looks like the update product hook might be called twice. Are there cases where `woocommerce_process_product_meta` would not be enough?

Also for the following line we actually only accept one argument:
```
add_action( 'woocommerce_process_product_meta', [ $this, 'handle_submission' ], 10, 2 );
```

nima-karimi committed:
Update channel visibility meta on process_product_meta action

nima-karimi committed:
Remove unnecessary arguments count

nima-karimi committed:
Check if product type is supported before updating visibility meta

nima-karimi commented:
> Two of those seem to be because it's hooking into both `woocommerce_update_product` and `woocommerce_process_product_meta`, looks like the update product hook might be called twice. Are there cases where `woocommerce_process_product_meta` would not be enough?

Good point. I didn't realize at the time that the `woocommerce_process_product_meta` itself is being called by the `save_post` action (couldn't find the usage since it was named with variables `'woocommerce_process_' . $post->post_type . '_meta'`). Seems like it's always called when a product is modified or created so there's no need for other actions.

I've removed the other actions and just left the `woocommerce_process_product_meta` to handle the meta box data. I also added a check for product type in there so that we don't store the visibility value for unsupported products.

Ready for another review @mikkamp

mikkamp reviewed (approved):
Looks good, although I'm not sure about the check for product type. At the time the hook is called the product hasn't been updated in the DB yet. So we are comparing the product type with the previous value, which will work in most cases but you can get around it in a scenario like this:

- Save product as "simple" and set visibility to sync-and-show
- Switch to "grouped" and update product
- When the `handle_submission` function is called the product type is still "simple"
- Product is saved as "grouped" with visibility set to sync-and-show

When we query products we only query valid product types, so it doesn't ever lead to any issues. Maybe we just need a simple note clarifying what's happening so we don't trip up over this in the future.
Or possibly get the product type from the `$_POST` value, although that would still leave the previous meta data intact so you'd be accomplishing the same as it is now.

nima-karimi commented:
Hmm, what if we go back to using the `woocommerce_new_product` and `woocommerce_update_product` actions? This way we have the final product object and it's already in the DB. We probably won't need to hook into the `woocommerce_process_product_meta` action if we hook into those two.

nima-karimi commented:
> * Save product as "simple" and set visibility to sync-and-show
> 
> * Switch to "grouped" and update product
> 
> * When the `handle_submission` function is called the product type is still "simple"
> 
> * Product is saved as "grouped" with visibility set to sync-and-show

I think this issue will always be there because we don't delete the visibility option if they change the product type. We could do that but it sounds unnecessary given that we double check the product type before syncing it.

nima-karimi committed:
Use `woocommerce_new_product` and `woocommerce_update_product` actions

Also introduces a static variable to check and prevent double submission.

nima-karimi commented:
I modified the filter to use `woocommerce_new_product` and `woocommerce_update_product` and also added a static variable to prevent double submission.

(Turns out that the `woocommerce_update_product` is called twice because of the `WC_Meta_Box_Product_Images::save` method which saves the product again in the same request.)

ref:  p1624278374114700-slack-C01E9LB4X61

nima-karimi committed:
Check for `isset` instead of `empty`

mikkamp reviewed (approved):
Thanks for all the refactoring. It works good and it's triggering the save only once when it's a valid product type.
